Add PLEP 8 on switching to calendar versioning (#39)
PlasmaPy/PlasmaPy-PLEPs: PLEP-0008.rst

  number.

* The quantity of changes to PlasmaPy is roughly proportional to the
  time between releases.

* The versions would be more consistent with projects like xarray,
  dask, pip, Ubuntu, and PyCharm, which each use some form of calendar
  versioning.

* The differences between the last two releases beginning with ``0.``
  are likely to be comparable to the differences between the last
  ``0.`` release and ``1.0.0``. Hence, a change to ``1.0.0`` would
  indicate more changes than were actually happening.

The disadvantages to adopting calendar versioning are:

* A change to calendar versioning would likely need to be permanent in
  order to preserve monotonicity.

* Calendar versioning is not used by many of the packages we depend on,
  like NumPy, SciPy, matplotlib, and Astropy.

* The version number itself would not indicate how mature the package
  is.

* Calendar-based version numbers of PlasmaPy would differ
  substantially from the version numbers of prior releases.

Versioning Specification
------------------------

PlasmaPy final releases will use `calendar based versioning`_ of the
form ``YY.0M.MICRO``, where:

> * The major release number ``YY`` is the year minus 2000,

Author: Would this be better with a two-digit year or a four-digit year?

Reviewer: 4-digit YYYY, for the same reasons as https://github.com/dask/community/issues/100#issuecomment-704499489 :)

Changes to the API (including backward incompatible changes) may only
occur when there is a change in the major and/or minor release
number. No changes to the API shall be made in bugfix releases.

Release schedule
----------------

Major/minor releases should be made every ∼4 months.

Alternatives
------------

This PLEP supersedes `PLEP 5`_, which is no longer in effect. PLEP 5
specified that PlasmaPy use `semantic versioning`_. In `semantic
versioning`_, versions are of the form ``MAJOR.MINOR.PATCH``.  For
releases following ``1.0.0``, ``MAJOR`` is incremented when there are
backwards incompatible changes to the API, ``MINOR`` is incremented
when there are backwards compatible changes, and ``PATCH`` is
incremented when the API does not change.  The main advantage of
semantic versioning is that it is easy to tell when there are
backwards incompatible changes. Because a community software project
will have features at different stages of development, it is likely
that there will be some backward incompatible changes during every
release. Hence, date-based versioning is more appropriate to PlasmaPy
than versioning based on whether or not backwards incompatible changes

> have been made.

Reviewer: "Because a community software project will have features"... ? I think I get what you were going for here, but can you rephrase?

  to drop support for prior releases of packages, in particular if there
  is a policy to support all releases made within a certain amount of
  time in the past (e.g., `NumPy Enhancement Proposal 29`_).

Disadvantages of adopting calendar versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Calendar versioning is not used by most of the packages PlasmaPy
  depends on (i.e., NumPy, SciPy, matplotlib, and Astropy).

* The version number itself would not indicate the maturity of the
  package.

* Calendar version numbers of PlasmaPy would differ substantially from
  the version numbers of prior releases.

* Changes to the version number would not indicate whether a release
  contains backward incompatible changes, as in `semantic versioning`_.

* The version number itself would not indicate whether a release is a
  long-term support (LTS) release.

* Versions of the form ``0.*.*`` (as currently in use) indicate that
  breaking changes could occur for any release.

Additional considerations for adopting calendar versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Calendar versioning is compatible with :pep:`440` and
  ``setuptools_scm``.

> * Using the full year in the version number will reduce the chance of
>   ambiguity. Similarly, using the zero-padded month will increase the
>   likelihood that the minor version number is interpreted as a month.

Reviewer 2: ```suggestion
* Using the full year in the version number will reduce the chance of
  ambiguity. Similarly.
```

I'm just deleting this, but feel free to add back a comment about the month but without the "zero-padded" statement.

PLEP-0008 – Calendar Versioning
===============================

+-------------------+---------------------------------------------+

> | PLEP              | number                                      |

Reviewer 2: ```suggestion
| PLEP              | 8                                           |
```

+===================+=============================================+
| title             | PlasmaPy Versioning and Releases            |
+-------------------+---------------------------------------------+
| author(s)         | Nick Murphy                                 |
+-------------------+---------------------------------------------+
| contact email     | namurphy@cfa.harvard.edu                    |
+-------------------+---------------------------------------------+
| date created      | 2022-11-22                                  |
+-------------------+---------------------------------------------+

> | date last revised | 2022-11-29                                  |

Reviewer 2: ```suggestion
| date last revised | 2022-12-08                                  |
```

[optional ... ?] just update to current date

+-------------------+---------------------------------------------+
| type              | process                                     |
+-------------------+---------------------------------------------+
| status            | discussion                                  |
+-------------------+---------------------------------------------+

> | DOI               |                                             |

Reviewer 2: update with reserved DOI

|                   |                                             |
+-------------------+---------------------------------------------+

Brief Description
-----------------

PlasmaPy will use `calendar versioning`_ starting with the first release
of 2023.

.. note::

    This PLEP supersedes |PLEP 5|, which is no longer in effect.

Versioning Specification
------------------------

PlasmaPy final releases will use `calendar versioning`_ of the

> form ``YYYY.0M.MICRO``, where:

Reviewer 2: Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

> form ``YYYY.0M.MICRO``, where:

Reviewer 2: ```suggestion
form ``YYYY.M.MICRO``, where:
```

Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

* The major release number ``YYYY`` is the year,

> * The minor release number ``0M`` is the zero-padded month, and

Reviewer 2: ```suggestion
* The minor release number ``M`` is positive integer representing the release month, and
```

See above PEP440 comment.  Feel to re-word to your own liking.

* The bugfix release number ``MICRO`` will start from ``0`` and be
  incremented by one for each bugfix release.

> For example, the first release in April 2023 would be ``2023.04.0``,
> with ``2023.04.1`` as the first bugfix release of the ``2023.04.*``
> series.

Reviewer 2: ```suggestion
For example, the first release in April 2023 would be ``2023.4.0``,
with ``2023.4.1`` as the first bugfix release of the ``2023.4.*``
series.
```

Again, refer to the above PEP440 comments.

Bugfix releases shall contain no changes to the API.

Tradeoff Analysis
-----------------

Advantages of adopting calendar versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> * The age of a release can be gauged directly from the version number.

Reviewer 2: Not sure the best way to make the comment, but also easier to traceback the version when cited in a journal article.

* The extent of changes between releases of a package tends to be
  roughly proportional to the time between releases, and can therefore
  be gauged more accurately with date-based versioning.

> * The versioning scheme would be consistent with projects that use some
>   form of calendar versioning such as xarray, dask, pip, Ubuntu, and
>   PyCharm.

Reviewer 2: ```suggestion
* The versioning scheme would be consistent with projects that use some
  form of calendar versioning such as ``xarray``, ``dask``, ``pip``,
  Ubuntu, and PyCharm.
```

[optional] ... but thank you for using the Oxford comma. 😺

  calendar versioning.

* Calendar versioning makes it easier to determine the appropriate time
  to drop support for prior releases of packages, in particular if there
  is a policy to support all releases made within a certain amount of
  time in the past (e.g., `NumPy Enhancement Proposal 29`_).

Disadvantages of adopting calendar versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Calendar versioning is not used by most of the packages PlasmaPy
  depends on (i.e., NumPy, SciPy, matplotlib, and Astropy).

* The version number itself would not indicate the maturity of the
  package.

* Calendar version numbers of PlasmaPy would differ substantially from
  the version numbers of prior releases.

* Changes to the version number would not indicate whether a release
  contains backward incompatible changes, as in `semantic versioning`_.

* The version number itself would not indicate whether a release is a
  long-term support (LTS) release.

* Versions of the form ``0.*.*`` (as currently in use) indicate that
  breaking changes could occur for any release.

Additional considerations for adopting calendar versioning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 
> * A change to calendar versioning would need to be permanent in order to
>   preserve version monotonicity.

Author: ```suggestion
```
There's something in PEP 440 about defining epochs, so this change need not be permanent.

------------------------

PlasmaPy final releases will use `calendar versioning`_ of the
form ``YYYY.MM.MICRO``, where:

* The major release number ``YYYY`` is the year,

> * The minor release number ``MM`` is positive integer representing the release month, and

Reviewer: ```suggestion
* The minor release number ``MM`` is a positive integer representing the release month, and
```
